cgapi.py

```diff
@@ -44,105 +120,195 @@ def __init__(self, response):
-        r = requests.post(url, headers=self.get_default_headers(), data=buf)
+        url = self.routes.get_file(submission_id=submission_id, path=path)
+        r = self.s.post(url, data=buf)
 
-        if r.status_code >= 400:
-            raise CGAPIException(r)
+        self._handle_response_error(r)
 
         return r.json()
 
     def rename_file(self, file_id, new_path):
-        url = APIRoutes.FILE_RENAME.format(file_id=file_id, new_path=new_path)
-        r = requests.patch(url, headers=self.get_default_headers())
+        url = self.routes.get_file_rename(file_id=file_id, new_path=new_path)
+        r = self.s.patch(url)
 
-        if r.status_code >= 400:
-            raise CGAPIException(r)
+        self._handle_response_error(r)
 
         return r.json()
 
     def get_file(self, file_id):
-        url = APIRoutes.FILE_BUF.format(file_id=file_id)
-        r = requests.get(url, headers=self.get_default_headers())
+        url = self.routes.get_file_buf(file_id=file_id)
+        r = self.s.get(url)
 
-        if r.status_code >= 400:
-            raise CGAPIException(r)
+        self._handle_response_error(r)
 
         return r.content
 
     def patch_file(self, file_id, buf):
-        url = APIRoutes.FILE_BUF.format(file_id=file_id)
-        r = requests.patch(url, headers=self.get_default_headers(), data=buf)
+        url = self.routes.get_file_buf(file_id=file_id)
+        r = self.s.patch(url, data=buf)
 
-        if r.status_code >= 400:
-            raise CGAPIException(r)
+        self._handle_response_error(r)
 
         return r.json()
 
     def delete_file(self, file_id):
-        url = APIRoutes.FILE_BUF.format(file_id=file_id)
-        r = requests.delete(url, headers=self.get_default_headers())
+        url = self.routes.get_file_buf(file_id=file_id)
+        r = self.s.delete(url)
+
+        self._handle_response_error(r)
+
+    def get_assignment_rubric(self, assignment_id):
+        url = self.routes.get_assignment_rubric(assignment_id)
+        r = self.s.get(url)
+        if r.status_code == 404:
+            return []
+
+        self._handle_response_error(r)
+
+        return r.json()
+
+    def set_assignment_rubric(self, assignment_id, rub):
+        url = self.routes.get_assignment_rubric(assignment_id)
+        r = self.s.put(url, json=rub)
+        self._handle_response_error(r)
+
+    def get_submission_rubric(self, submission_id):
+        url = self.routes.get_submission_rubric(submission_id)
+        r = self.s.get(url)
+
+        if r.status_code == 404:
+            return {'rubrics': [], 'selected': []}
+
+        self._handle_response_error(r)
+
+        return r.json()
+
+    def select_rubricitems(self, submission_id, items):
+        url = self.routes.select_rubricitems(submission_id)
+        r = self.s.patch(url, json={'items': items})
+        self._handle_response_error(r)
+
+    def get_feedbacks(self, assignment_id):
```

Comment:
Maybe we should call this `get_line_feedback` or something, to better differentiate it from `get_feedback`? And also because "feedbacks" isn't really a word.

Reply:
WRONG, feedbacks is a word.

Reply:
Hmm I found some sources staying that it has indeed become acceptable to use feedbacks as a plural form of feedback, so fair enough. But still it sounds a bit awkward and I'd still appreciate some differentiation with `get_feedback`. As it is now I would expect this function to do the equivalent of multiple calls to that function.

Reply:
Oh wow, I was just making a joke. A different name would be fine by me :)

Reply:
LOL. You were correct though, it seems